### ISSUE TITLE
Delay ADS stream until XDS cluster warming

### ISSUE THREAD

**Issue by snowp** [envoyproxy/envoy#7253]
Currently the ADS stream is started immediately after static load completes:

https://github.com/envoyproxy/envoy/blob/master/source/common/upstream/cluster_manager_impl.cc#L283

When used with STRICT_DNS, this means that the gRPC stream keeps attempting to connect until DNS resolves, which can exacerbate startup delays due to DNS issues: the stream will continuously attempt to connect, building up the retry backoff counter. By the time DNS finally resolves, the next attempt might be scheduled several seconds into the future. As a result, a 5 second DNS delay (eg due to a timeout) might result in delaying startup by >10s.

While this slowdown isn't a major issue, it's caused short-ish DNS issues to cause significant deployment delays for us.

Conceptually the fix sounds like it would be to defer ADS stream start until the associated cluster warms (eg wait for first DNS resolution), but I'm not sure how easy it would be to express that with the current init helper.

**Comment by htuch** [comment]
@snowp I think you could potentially take advantage of https://github.com/envoyproxy/envoy/blob/master/source/common/upstream/cluster_manager_impl.cc#L208. Shouldn't be too hard I think.

**Comment by htuch** [comment]
@snowp are you interested in implementing or should we mark as help wanted? It seems a reasonable request.

**Comment by snowp** [comment]
Yeah I have a half done branch, I'll try to find the time to put it up

**Comment by snowp** [comment]
Marking help wanted since I don't seem to be able to find the time to work on this

**Comment by htuch** [comment]
@yanavlasov FYI; not related to your current PR but some context on earlier discussions around ADS and warming.
